### PR TITLE
Metadata: Add Fluent API to configure Indexed properties

### LINKED ISSUE #13610: Indexed properties mapping support (part of property bag entities)
From discussion with @divega and @ajcvickers. We would like to support Entity classes with indexers:

```csharp
    public class SomeEntity
    {
        public object this[string propertyName]
        {
            get
            {
                // get value of property named propertyName
            }

            set
            {
                // set value of property named propertyName
            }
        }
    }
```

You can then declare an indexed property in `OnModelCreating()` and have EF use the indexer to access that property, instead of expecting a member. (Note: this addresses parts of issues #2968 and #9914).

Previous Tasks:
- [ ] Update the metadata to allow the model to get and set property values using the indexer
- [ ] Store the `PropertyInfo` in `PropertyBase`
- [ ] Add `ModelBuilder` fluent API so the user does not need to access metadata methods directly
- [ ] Update query pipeline to allow indexer references e.g. `ctx.Entity<SomeEntity>.Property(e => e["PropertyName"])...`

New Tasks:
- [x] Update metadata to allow model to get/set property values using the indexer

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #13610

```C#
modelBuilder.Entity<City>().IndexedProperty<string>("Nation");
```

### CODE REVIEW

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs [hunk 1]
```diff
@@ -180,6 +180,36 @@ public virtual PropertyBuilder Property([NotNull] Type propertyType, [NotNull] s
...
+                Builder.IndexedProperty(
+                    typeof(TProperty),
+                    Check.NotEmpty(propertyName, nameof(propertyName)), ConfigurationSource.Explicit).Metadata);
+
+        /// <summary>
+        ///     <para>
+        ///         Returns an object that can be used to configure a property of the entity type.
+        ///         If no property with the given name exists, then a new property will be added.
+        ///     </para>
+        /// </summary>
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: Add a paragraph to these docs briefly explaining indexed properties

Comment (smitpatel) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: Can you suggest a para?

File: src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 2]
```diff
@@ -597,7 +619,11 @@ public virtual InternalPropertyBuilder Property([NotNull] MemberInfo memberInfo,
                     }
 
                     property = clrProperty != null
-                        ? Metadata.AddProperty(clrProperty, configurationSource.Value)
+                        ? (clrProperty is PropertyInfo propertyInfo
+                            && propertyInfo == Metadata.FindIndexerPropertyInfo()
+                            ? Metadata.AddProperty(
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs]: You can call the same overload in both cases

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: Indexed properties are stored in the entity using [an index](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/indexers/) suppling the provided property name.

Comment (smitpatel) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: How to add hyper link in XML docs?

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: Dunno. @bricelam ?

Comment (bricelam) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: `...using <see href="https://docs.microsoft.com/dotnet/csharp/programming-guide/indexers/">an indexer</see> supplying...`

Comment (bricelam) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: https://dotnet.github.io/docfx/spec/triple_slash_comments_spec.html

Comment (smitpatel) [reply src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: Adding it in my next PR.

**Pull request merged**